File: globalpayments-android-sdk/src/main/java/com/globalpayments/android/sdk/TaskExecutor.java
package com.globalpayments.android.sdk;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class TaskExecutor {
    private final Executor executor = Executors.newSingleThreadExecutor();
    private final Handler mainThreadHandler = new Handler(Looper.getMainLooper());

    private <T> void execute(final Task<T> task) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    final T result = task.executeAsync();
                    if (result == null) {
                        callOnError(task, new Exception("Null response"));
                    } else {
                        callOnSuccess(task, result);
                    }
                } catch (final Exception e) {
                    callOnError(task, e);
                }
            }
        });
    }

    private <T> void callOnSuccess(final Task<T> task, final T result) {
        mainThreadHandler.post(new Runnable() {
            @Override
            public void run() {
                task.onSuccess(result);
            }
        });
    }

    private <T> void callOnError(final Task<T> task, final Exception exception) {
        mainThreadHandler.post(new Runnable() {
            @Override
            public void run() {
                task.onError(exception);
            }
        });
    }

    public static <T> void executeAsync(final Task<T> task) {
        new TaskExecutor().execute(task);
    }

    public interface Task<T> {
        T executeAsync() throws Exception;
        void onSuccess(T value);
        void onError(Exception exception);
    }
}

File: globalpayments-android-sdk/src/main/java/com/globalpayments/android/sdk/task/TaskExecutor.kt
package com.globalpayments.android.sdk.task

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executors
import java.util.concurrent.Future

object TaskExecutor {

    private val mainHandler = Handler(Looper.getMainLooper())
    private val executorService =
        Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() / 2)

    /**
     * Execute a method from sdk on a background thread.
     * The result is returned on the same thread on which is executed. This is done to avoid context switching when we want to chain multiple calls.
     * @param shouldReturnToMainThread boolean that tells if the onFinished callback should be called on the main looper
     * @param taskToExecute the task we want the execute
     * @param onFinished callback on which the result is returned
     *
     * @return the return value is a future used for cancellation
     */
    fun <T> executeAsync(
        shouldReturnToMainThread: Boolean = true,
        taskToExecute: () -> T?,
        onFinished: ((TaskResult<T>) -> Unit)? = null
    ): Future<*> {
        return executorService.submit {
            val result: TaskResult<T> = try {
                val result = taskToExecute()
                if (result == null) {
                    TaskResult.Error(RuntimeException("Null result"))
                } else {
                    TaskResult.Success(result)
                }
            } catch (exception: Exception) {
                TaskResult.Error(exception)
            }
            if (shouldReturnToMainThread) {
                mainHandler.post { onFinished?.invoke(result) }
                return@submit
            }
            onFinished?.invoke(result)
        }
    }

}